posts/simp-made-simple.md

```diff
@@ -0,0 +1,330 @@
+
+## The `SimpM` monad
+
+In this section, we take a look at another key component of the internals of simp, namely the `SimpM` monad. 
+
+### `SimpM`
+
+[`SimpM`](https://leanprover-community.github.io/mathlib4_docs/find/?pattern=Lean.Meta.Simp.SimpM#doc) is the monad that tracks the current context `simp` is running in (what `simp` theorems are available, etc) and what has been done so far (e.g. number of steps so far, theorems used).
+In particular this also captures the `MetaM` context.
+
+Let's go through this in more detail. The monad `SimpM` is defined using monad transformers as follows:
+```lean
+abbrev SimpM := ReaderT Simp.MethodsRef $ ReaderT Simp.Context $ StateRefT Simp.State MetaM
+```
+
+Let's go through these steps one by one.
+
+1) The monad `MetaM`. This is one of the fundamental monads for metaprogramming in Lean. 
+  The state of `MetaM` allows one to access things like:
+    - Information about the file we're running in (e.g. name, imports, etc)
+    - Information about what definitions/theorems we're allowed to use
+    - What local variables/declarations we have access to
+
+2) The first monad transformer application: `StateRefT Simp.State MetaM`. 
+  The idea here is the following: since the goal of the `SimpM` monad is to track the state of a `simp` call
+  (i.e. what's happening, as the program runs), we need to capture more information than what `MetaM` gives us. 
+  Specifially, we want a monad that can track the state of what's happening via the following structure: 
+  ```lean
+  structure Simp.State where
+    cache        : Cache
+    congrCache   : CongrCache
+    dsimpCache   : ExprStructMap Expr
+    usedTheorems : UsedSimps
+    numSteps     : Nat
+    diag         : Diagnostics
+  ```
+  This is something we can achieve using the `StateRefT` monad transformer, which takes as input a state type (`Simp.State` in our case) and a monad, and creates a new monad that can read _and write_ this state. In other words, `StateRefT Simp.State MetaM` is a souped up version of `MetaM` that can now track extra information by storing (and updating) at term of type `Simp.State`.
+
+3) The second monad transformer application: `ReaderT Simp.Context $ StateRefT Simp.State MetaM`. 
+  The `SimpM` monad should also be able to access the "context" that `simp` is running in, e.g. which simp theorems it has access to and so on. This is captured by the type `Simp.Context`.
+  Here, the situation is not quite the same as when we were adding a `Simp.State` state to `MetaM`: while we will often want to change the state during the `simp` call, the context should always be the same (in programmer lingo: _immutable_)
+  Thus, we use a different monad transformer called `ReaderT`, which is almost identical to `StateT`, but outputs a new monad that can only read the type passed as parameter.
```

Comment:
Mildly pedantic, but `ReaderT` can change the value in the calls it makes (e.g. [`ReaderT.adapt`](https://leanprover-community.github.io/mathlib4_docs/Init/Prelude.html#ReaderT.adapt) as equivalent to `modify` for a Reader). In programming language nerd terms, `State(Ref)T` gives you a global variable, and `ReaderT` gives you dynamic scoping.

(Does simp in practice use that though?)